src/submodule.c

```diff
@@ -1915,32 +1919,34 @@ static int submodule_load_from_wd_lite(git_submodule *sm)
 }
 
 /**
- * Returns a snapshot of $WORK_TREE/.gitmodules.
+ * Requests a snapshot of $WORK_TREE/.gitmodules.
  *
- * We ignore any errors and just pretend the file isn't there.
+ * Returns GIT_ENOTFOUND in case no .gitmodules file exist
  */
-static git_config *gitmodules_snapshot(git_repository *repo)
+static int gitmodules_snapshot(git_config **snap, git_repository *repo)
 {
 	const char *workdir = git_repository_workdir(repo);
-	git_config *mods = NULL, *snap = NULL;
+	git_config *mods = NULL;
 	git_buf path = GIT_BUF_INIT;
+	int error;
 
-	if (workdir != NULL) {
-		if (git_buf_joinpath(&path, workdir, GIT_MODULES_FILE) != 0)
-			return NULL;
+	if (!workdir)
+		return GIT_ENOTFOUND;
 
-		if (git_config_open_ondisk(&mods, path.ptr) < 0)
-			mods = NULL;
-	}
+	if (git_buf_joinpath(&path, workdir, GIT_MODULES_FILE) != 0)
+		return -1;
+
+	error = git_config_open_ondisk(&mods, path.ptr);
```

Comment:
You should probably also check the return value here. While an error will cause `mods` to be `NULL`, which is catched in a few lines, it feels a bit more robust to check directly. You could avoid multiple free's by adding a `goto out`. Additional kudos if you also convert the other error handling branches to `goto out` with a unified section for freeing up memory in case `error != 0`.